### PR TITLE
8329982: compiler/jvmci/jdk.vm.ci.code.test/src/jdk/vm/ci/code/test/SimpleDebugInfoTest.java failed assert(oopDesc::is_oop_or_null(val)) failed: bad oop found

### PR DESCRIPTION
**Pull request by dougxc**
This PR adds the missing nmethod entry barriers to JVMCI hand assembled tests.
It also closes the escape hatch in jvmciCodeInstaller.cpp that allowed JVMCI code to be installed without nmethod entry barriers.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329982](https://bugs.openjdk.org/browse/JDK-8329982): compiler/jvmci/jdk.vm.ci.code.test/src/jdk/vm/ci/code/test/SimpleDebugInfoTest.java failed assert(oopDesc::is_oop_or_null(val)) failed: bad oop found (**Bug** - P3)

### Reviewers
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19035/head:pull/19035` \
`$ git checkout pull/19035`

Update a local copy of the PR: \
`$ git checkout pull/19035` \
`$ git pull https://git.openjdk.org/jdk.git pull/19035/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19035`

View PR using the GUI difftool: \
`$ git pr show -t 19035`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19035.diff">https://git.openjdk.org/jdk/pull/19035.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19035#issuecomment-2088605549)

### CODE REVIEW

File: src/hotspot/share/jvmci/jvmciCodeInstaller.cpp [hunk 1]
```diff
@@ -775,7 +775,7 @@ JVMCI::CodeInstallResult CodeInstaller::install(JVMCICompiler* compiler,
     // configurations which generate assembly without being a full compiler. So for now we enforce
     // that JIT compiled methods must have an nmethod barrier.
     bool install_default = JVMCIENV->get_HotSpotNmethod_isDefault(installed_code) != 0;
```
Comment (tkrodriguez) [inline src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:777]: This line is no longer needed.

**Comment by tkrodriguez** [comment]
In the long term I'm not sure it's worth trying to maintain these assembler tests.  The barrier verification code is very weak and on aarch64 it's slightly complicated so we're barely checking that it really matches.  I guess this is good enough until we get further problems.

I think you can simplify some other logic that deals with the optionality of the barrier.  Start with removing JVMCINMethodData::has_entry_barrier and maybe update some of the comments to reflect that it's always emitted.  And places that check _nmethod_entry_patch_offset for -1 can be removed or weakened.

**Comment by dougxc** [comment]
> In the long term I'm not sure it's worth trying to maintain these assembler tests. The barrier verification code is very weak and on aarch64 it's slightly complicated so we're barely checking that it really matches. I guess this is good enough until we get further problems.

I agree. I had to push more changes now to remove tests that expect to be able to install 0 length code (which obviously fail the nmethod barrier verification). These tests provided stop gap coverage in the early days of JVMCI but now test functionality where breakage will clearly show up in higher layers (such as Graal). What's more, expanding the assembler support in JVMCI is redundant with the fully fledged assembler in Graal.

> I think you can simplify some other logic that deals with the optionality of the barrier. Start with removing JVMCINMethodData::has_entry_barrier and maybe update some of the comments to reflect that it's always emitted. And places that check _nmethod_entry_patch_offset for -1 can be removed or weakened.

I've done that now. Please let me know if you can spot any other vestiges of the optional support.

**Comment by tkrodriguez** [comment]
It would be super nice if we could figure out a clean way to share canned snippets of assembly from HotSpot back through JVMCI.  There are lots of potential complexities though:  register usage, the jcc erratum, relocations, fast/slow splits.  The emit function could be called from the Graal assembler so that the sizing and alignment can be properly handled.  HotSpot relocations could be translated in some fashion and maybe labels could be handled as well.

The nmethod entry barrier fast path emission could probably be handled fairly cleanly since it's mostly a straightline snippet with a conditional branch at the end.  It's just unclear if building that machinery is more complicated than maintaining and checking a clone of a small piece of assembly.  The TestAssembler is a dubious piece of code given the complexity of emitting real nmethods.  It doesn't even support the complex return sequence being used these days.

**Comment by dougxc** [comment]
> The TestAssembler is a dubious piece of code given the complexity of emitting real nmethods. It doesn't even support the complex return sequence being used these days.

The next time there's a TestAssembler failure due to a change in nmethod invariants, I will remove it completely. There is sufficient coverage now in Graal tests that it no longer offers sufficient value.

**Review by tkrodriguez — APPROVED** [review]
Sounds and looks good.

**Comment by dougxc** [comment]
Thanks for the feedback and reviews.

/integrate

**Comment by dougxc** [comment]
> The next time there's a TestAssembler failure due to a change in nmethod invariants, I will remove it completely. There is sufficient coverage now in Graal tests that it no longer offers sufficient value.

Happened again: https://bugs.openjdk.org/browse/JDK-8336242

I'll put up a PR to remove the problematic test.

**Pull request closed without merging**